go/pkg/cas/upload.go

```diff
@@ -358,24 +388,69 @@ func (u *uploader) visitDir(ctx context.Context, absPath string) (*repb.Director
 
 	wgChildren.Wait()
 	if subErr != nil {
-		return nil, errors.Wrapf(subErr, "failed to read the directory %q entirely", absPath)
+		return nil, errors.Wrapf(subErr, "failed to read the directory %q entirely", key.AbsPath)
 	}
 
-	item := uploadItemFromDirMsg(absPath, dir)
+	item := uploadItemFromDirMsg(key.AbsPath, dir)
 	if err := u.scheduleCheck(ctx, item); err != nil {
 		return nil, err
 	}
 	return &repb.DirectoryNode{
-		Name:   filepath.Base(absPath),
+		Name:   filepath.Base(key.AbsPath),
 		Digest: item.Digest,
 	}, nil
 }
 
-// visitSymlink converts a symlink to a SymlinkNode and schedules visitation
+// visitSymlink converts a symlink to a directory node and schedules visitation
 // of the target file.
-func (u *uploader) visitSymlink(ctx context.Context, absPath string) (*repb.SymlinkNode, error) {
-	// TODO(nodir): implement
-	panic("not supported")
+// If key.PreserveSymlinks is true, then returns a SymlinkNode, otherwise
+// returns the directory node of the target file.
+func (u *uploader) visitSymlink(ctx context.Context, key fsCacheKey) (proto.Message, error) {
+	target, err := os.Readlink(key.AbsPath)
+	if err != nil {
+		return nil, errors.Wrapf(err, "os.ReadLink")
+	}
+
+	// Determine absolute and relative paths of the target.
+	var absTarget, relTarget string
+	symlinkDir := filepath.Dir(key.AbsPath)
+	target = filepath.Clean(target) // target may end with slash
+	if filepath.IsAbs(target) {
+		absTarget = target
+		if relTarget, err = filepath.Rel(symlinkDir, absTarget); err != nil {
+			return nil, err
+		}
+	} else {
+		relTarget = target
+		absTarget = filepath.Join(symlinkDir, relTarget)
+	}
+	// TODO(nodir): add an option to return an error if a symlink taret outside of
+	// execution root is detected.
+
+	symlinkNode := &repb.SymlinkNode{
+		Name:   filepath.Base(key.AbsPath),
+		Target: filepath.ToSlash(relTarget),
+	}
+
+	targetInfo, err := os.Lstat(absTarget)
+	switch {
+	case os.IsNotExist(err) && key.PreserveSymlinks && key.AllowDanglingLinks:
+		// Special case for preserved dangling links.
+		return symlinkNode, nil
+	case err != nil:
+		return nil, errors.WithStack(err)
+	}
+
+	targetKey := key
+	targetKey.AbsPath = absTarget
+	if !key.PreserveSymlinks {
+		return u.visitFile(ctx, targetKey, targetInfo)
+	}
+
+	// Even though we return a SymlinkNode, we still must visit the target file
+	// to ensure it is uploaded.
+	_, err = u.visitFile(ctx, targetKey, targetInfo)
+	return symlinkNode, err
```

Comment:
```suggestion
      node, err := u.visitFile(ctx, targetKey, targetInfo)
	if key.PreserveSymlinks {
		return symlinkNode, err
	}
	return node, err
```
wdyt?

Reply:
this is exactly the code I had before, but I figured that a comment is warranted because otherwise it is super obvious why do read the node and throw it away. I'll commit your change and restore the comment.